compiler/resolution/functionResolution.cpp

```diff
@@ -8883,3 +8899,25 @@ setScalarPromotionType(AggregateType* ct) {
       ct->scalarPromotionType = field->type;
   }
 }
+
+//
+// Does this function call a ref-return function that relies on the
+// setter param?  By induction this handles the general recursive case.
+//
+static bool functionUsesSetter(FnSymbol* fn) {
+  std::vector<CallExpr*> calls;
+  bool                   retval = false;
+
+  collectCallExprs(fn, calls);
+
+  for (size_t i = 0; i < calls.size() && retval == false; i++) {
```

Comment:
On Jan 13, 2016, at 10:07 AM, Ben Harshbarger <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/resolution/functionResolution.cpp<https://github.com/chapel-lang/chapel/pull/3101#discussion_r49626528>:

> @@ -8883,3 +8899,25 @@ setScalarPromotionType(AggregateType* ct) {
>        ct->scalarPromotionType = field->type;
>    }
>  }
> +
> +//
> +// Does this function call a ref-return function that relies on the
> +// setter param?  By induction this handles the general recursive case.
> +//
> +static bool functionUsesSetter(FnSymbol* fn) {
> +  std::vector<CallExpr*> calls;
> +  bool                   retval = false;
> +
> +  collectCallExprs(fn, calls);
> +
> +  for (size_t i = 0; i < calls.size() && retval == false; i++) {


Would a "for_vector" macro with a "break" statement be better here? I think this could go in as-is, but doesn't match the usual compiler conventions.

—
Reply to this email directly or view it on GitHub<https://github.com/chapel-lang/chapel/pull/3101/files#r49626528>.

Thanks.  As I think you are acknowledging, this is a style issue.  I prefer to
avoid even structured gotos when practical and I find standard for loops
as easy to read as the various macro-based expressions.

Mike